design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,264 @@
+```
+
+Given a string, parses it as a JSON value and returns the length of the top-level array. Returns
+`NULL` if the string can't be interpreted as a JSON array, i.e., it is `NULL` or it does not contain
+valid JSON, or the JSON value is not an array.
+
+#### Examples
+
+```
+json_array_length("[1, 2, 3]") // returns 3
+json_array_length("[1, [1, [2]], 3]") // returns 3
+json_array_length("[]") // returns 0
+json_array_length("{}") // returns NULL
+json_array_length("123") // returns NULL
+json_array_length("abc") // returns NULL
+json_array_length(NULL) // returns NULL
+```
+
+### json_keys
+
+```
+json_keys(json_string) -> Array<String>
+```
+
+Given a string, parses it as a JSON object and returns a ksqlDB array of strings representing the
+top-level keys. Returns `NULL` if the string can't be interpreted as a JSON object, i.e., it is
+`NULL` or it does not contain valid JSON, or the JSON value is not an object.
+
+#### Examples
+
+```
+json_keys("{\"a\": \"abc\", \"b\": { \"c\": \"a\" }, \"d\": 1}") // returns ["a", "b", "d"]
+json_keys("{}") // returns []
+json_keys("[]") // returns NULL
+json_keys("") // returns NULL
+json_keys("123") // returns NULL
+json_keys("abc") // returns NULL
+json_keys(NULL) // returns NULL
+```
+
+### json_records
+
+```
+json_records(json_string) -> Array<Struct<json_key:String, json_value:String>>
+```
+
+Given a string, parses it as a JSON object and returns a ksqlDB array of structs containing 2 
+strings - `json_key` and `json_value` representing the top-level keys and values. Returns `NULL` if 
+the string can't be interpreted as a JSON object, i.e. it is `NULL` or it does not contain valid 
+JSON, or the JSON value is not an object.
+
+#### Examples
+
+```
+json_records("{\"a\": \"abc\", \"b\": { \"c\": \"a\" }, \"d\": 1}") // returns [Struct{json_key="a", json_value="\"abc\""}, Struct{json_key="b", json_value="{ \"c\": \"a\" }}, Struct{json_key="d", json_value="1"}]
+json_records("{}") // returns []
+json_records("[]") // returns NULL
+json_records("") // returns NULL
+json_records("123") // returns NULL
+json_records("abc") // returns NULL
+json_records(NULL) // returns NULL
+```
+
+#### Rejected alternatives
+
+**Remove redundant quotes around values**. For example:
+
+```
+json_records("{\"a\": \"abc\"}) // returns [["a", "abc"]]
+```
+
+This makes the output less awkward, but removes the type information in certain cases. In the following
+case it is not possible to derive whether `1` is a string or a number:
+
+```
+json_records("{\"a\": \"1\"}) // returns [Struct{json_key="a", json_value="1"}]
+```
+
+**Return array of arrays rather than an array of structs**. For example:
+```
+json_records("{\"a\": \"abc\"}) // returns [["a", "abc"]]
+```
+
+A struct is a better return type in this case as it enables meaningful field names and reduces
+the number of potential sanity checks on the end-user side (e.g., array length).
+
+### json_concat
+
+```
+json_concat(json_string, json_string) -> String
```

Comment:
(Aleks suggested I comment here.). Should/could this method take varargs?

As one thing to think about, should we have a UDAF to concat multiple entries?  (Which would/should likely be a separate function.)

Reply:
I think this is great idea. ksqlDB's UDFs support varargs, so we can just change this method to `json_concat(json_strings...)` without any downsides.

Reply:
If the UDF frameworks supports it, sure!